src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1522 @@
+            >>> f = R([123, 129, 63, 14, 51, 76, 133])
+            >>> g = R([106, 134, 32, 41, 158, 115, 115])
+            >>> f.divrem(g)
+            (21, 106*x^5 + 156*x^4 + 131*x^3 + 43*x^2 + 86*x + 16)
+        """
+        cdef fmpz_t f
+        cdef fmpz_mod_poly Q, R
+
+        other = self.ctx.any_as_fmpz_mod_poly(other)
+        if other is NotImplemented:
+            raise TypeError(f"Cannot interpret {other} as a polynomial")
+
+        Q = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        R = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        Q.ctx = self.ctx
+        R.ctx = self.ctx
+
+        fmpz_init(f)
+        fmpz_mod_poly_divrem_f(
+            f, Q.val, R.val, self.val, (<fmpz_mod_poly>other).val, self.ctx.mod.val
+        )
+        if not fmpz_is_one(f):
+            fmpz_clear(f)
+            raise ValueError(
+                f"Cannot compute divrem of {self} with {other}"
+            )
+
+        return Q, R
+
+    def gcd(self, other):
+        """
+        Return the greatest common divisor of self and other.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> x = R.gen()
+            >>> f = x*(x + 1)
+            >>> f.gcd(x+1)
+            x + 1
+            >>> f.gcd(x*x)
+            x
+
+        """
+        cdef fmpz_mod_poly res
+
+        if not self.ctx.is_prime():
+            raise NotImplementedError("gcd algorithm assumes that the modulus is prime")
+                 
+        other = self.ctx.any_as_fmpz_mod_poly(other)
+        if other is NotImplemented:
+            raise TypeError(f"Cannot interpret {other} as a polynomial")
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = self.ctx
+        fmpz_mod_poly_gcd(
+            res.val, self.val, (<fmpz_mod_poly>other).val, self.ctx.mod.val
+        )
+        return res
+
+    def xgcd(self, other):
+        r"""
+        Computes the extended gcd of self and other: (`G`, `S`, `T`)
+        where `G` is the ``gcd(self, other)`` and `S`, `T` are such that:
+
+        :math:`G = \textrm{self}*S +  \textrm{other}*T`
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> f = R([143, 19, 37, 138, 102, 127, 95])
+            >>> g = R([139, 9, 35, 154, 87, 120, 24])
+            >>> f.xgcd(g)
+            (x^3 + 128*x^2 + 123*x + 91, 17*x^2 + 49*x + 104, 21*x^2 + 5*x + 25)
+
+        """
+        cdef fmpz_mod_poly G, S, T
+        cdef fmpz_t f
+
+        other = self.ctx.any_as_fmpz_mod_poly(other)
+        if other is NotImplemented:
+            raise TypeError(f"Cannot interpret {other} as a polynomial")
+
+        G = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        S = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        T = fmpz_mod_poly.__new__(fmpz_mod_poly)
+
+        G.ctx = self.ctx
+        S.ctx = self.ctx
+        T.ctx = self.ctx
+
+        fmpz_init(f)
+        fmpz_mod_poly_xgcd_f(
+            f, G.val, S.val, T.val, self.val, (<fmpz_mod_poly>other).val, self.ctx.mod.val
+        )
```

Comment:
When exactly does any `fmpz_mod_poly_init_*` function get called in the above?

Reply:
Inside the `__new__()` call, thanks to the `__cinit__()` on the `fmpz_mod_poly` type?

Reply:
The `fmpz_mod_poly.__cinit__` method does not call any init function:
https://github.com/flintlib/python-flint/blob/69b2e50f4ff7bb5031e29054884818cc8071c19e/src/flint/types/fmpz_mod_poly.pyx#L328-L340
When `__new__` is called directly only `__cinit__` will be called (`__init__` is skipped). Otherwise when `__init__` is called it does not call any `fmpz_mod_init_*` function and instead calls `set_any_as_fmpz_mod_poly` which also does not call any `fmpz_mod_init_*` functions.

Reply:
oh you're correct... I must have never included it when we refactored to try and avoid the call when `ctx` wasn't available... Hmm

Reply:
Which leads me back to wanting to call `init` within `cinit`. but then I need to be careful that there's a context available...